ttnn/cpp/ttnn/operations/ccl/all_gather/device/all_gather_op.cpp

```diff
@@ -13,20 +13,73 @@
 
 namespace ttnn {
 
-AllGatherMode choose_all_gather_mode(Tensor const& input_tensor, Tensor const& output_tensor, uint32_t dim) {
-    bool is_sharded = input_tensor.is_sharded();
-
-    if (is_sharded) {
-        if (input_tensor.buffer()->shard_spec().tensor2d_shape[0] > 1) {
-            return AllGatherMode::FULL_WORKER_GRID_SHARDED;
-        } else {
-            return AllGatherMode::SINGLE_TILE_HIGH_WIDTH_SHARDED;
+
+AllGatherBidirectionalMode AllGatherConfig::choose_bidirectional_mode(Tensor const& input_tensor) {
+    std::size_t eth_l1_capacity = eth_l1_mem::address_map::MAX_L1_LOADING_SIZE - eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE;
+    std::size_t tensor_size_bytes = input_tensor.shape().volume() * input_tensor.element_size();
+    // This is currently a guestimate. We need a lot more hard data to identify where this dividing line is.
+    bool perf_degradation_from_full_tensor_mode = tensor_size_bytes > (2 * eth_l1_capacity);
+    if (perf_degradation_from_full_tensor_mode) {
+        return AllGatherBidirectionalMode::SPLIT_TENSOR;
+    }
+    return AllGatherBidirectionalMode::FULL_TENSOR;
+}
+
+AllGatherConfig::AllGatherConfig(Tensor const& input_tensor, Tensor const& output_tensor, uint32_t dim, uint32_t ring_size, uint32_t num_links, all_gather_op::Topology topology) :
+    num_links(num_links),
+    semaphore_size(32),
+    ring_size(ring_size),
+
+    erisc_handshake_address(tt::round_up(eth_l1_mem::address_map::ERISC_L1_UNRESERVED_BASE, 16)),
```

Comment:
16, is there an L1_ALIGNMENT define you can use?

Reply:
This is old code (just moved) but ideally there is something exposed at the arch level that gives us eth word alignment size. Is there such a variable?

Reply:
Opened this issue here to track the addition of some of these constants:
https://github.com/tenstorrent/tt-metal/issues/11156